### PR TITLE
Make get_daemons_stats_socket function asynchronous

### PR DESCRIPTION
**Pull request by mhamra**
|Related issue|
|---|
|#19394|

## Description
Make the get_daemons_stats_socket function asynchronous

## Tests

- [ ] Unit tests
- [ ] Api Integration Tests

### CODE REVIEW

**Comment by mhamra** [comment]
### API INTEGRATION TEST FAILURES

After running the AIT in a local environment with branch master and enhancement/19394-async-daemons-stats, I've found no difference in the failures, concluding that the failures were not introduced by the changes made on this PR.



```
-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
============================================================================================ short test summary info ============================================================================================
FAILED test_rbac_white_cluster_endpoints.tavern.yaml::/cluster/local/config
FAILED test_rbac_white_cluster_endpoints.tavern.yaml::GET /cluster/healthcheck
FAILED test_rbac_white_cluster_endpoints.tavern.yaml::GET /cluster/ruleset/synchronization
FAILED test_rbac_white_cluster_endpoints.tavern.yaml::GET /cluster/local/info
FAILED test_rbac_white_cluster_endpoints.tavern.yaml::GET /cluster/nodes
FAILED test_rbac_white_cluster_endpoints.tavern.yaml::GET /cluster/nodes/{node_id}
FAILED test_rbac_white_cluster_endpoints.tavern.yaml::GET /cluster/status
FAILED test_rbac_white_cluster_endpoints.tavern.yaml::GET /cluster/{node_id}/configuration
FAILED test_rbac_white_cluster_endpoints.tavern.yaml::GET /cluster/configuration/validation
FAILED test_rbac_white_cluster_endpoints.tavern.yaml::GET /cluster/{node_id}/configuration/{component}/{configuration}
FAILED test_rbac_white_cluster_endpoints.tavern.yaml::GET /cluster/{node_id}/info
FAILED test_rbac_white_cluster_endpoints.tavern.yaml::GET /cluster/{node_id}/logs
FAILED test_rbac_white_cluster_endpoints.tavern.yaml::GET /cluster/{node_id}/logs/summary
FAILED test_rbac_white_cluster_endpoints.tavern.yaml::GET /cluster/{node_id}/daemons/stats
FAILED test_rbac_white_cluster_endpoints.tavern.yaml::GET /cluster/{node_id}/stats
FAILED test_rbac_white_cluster_endpoints.tavern.yaml::GET /cluster/{node_id}/status
FAILED test_rbac_white_cluster_endpoints.tavern.yaml::GET /cluster/api/config
FAILED test_rbac_white_cluster_endpoints.tavern.yaml::PUT /cluster/restart
FAILED test_rbac_white_cluster_endpoints.tavern.yaml::PUT /cluster/{node_id}/configuration
============================================================================ 19 failed, 25 passed, 22 warnings in 732.83s (0:12:12) =============================================================================
```

```
-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
============================================================================================ short test summary info ============================================================================================
FAILED test_rbac_white_cluster_endpoints.tavern.yaml::/cluster/local/config
FAILED test_rbac_white_cluster_endpoints.tavern.yaml::GET /cluster/healthcheck
FAILED test_rbac_white_cluster_endpoints.tavern.yaml::GET /cluster/ruleset/synchronization
FAILED test_rbac_white_cluster_endpoints.tavern.yaml::GET /cluster/local/info
FAILED test_rbac_white_cluster_endpoints.tavern.yaml::GET /cluster/nodes
FAILED test_rbac_white_cluster_endpoints.tavern.yaml::GET /cluster/nodes/{node_id}
FAILED test_rbac_white_cluster_endpoints.tavern.yaml::GET /cluster/status
FAILED test_rbac_white_cluster_endpoints.tavern.yaml::GET /cluster/{node_id}/configuration
FAILED test_rbac_white_cluster_endpoints.tavern.yaml::GET /cluster/configuration/validation
FAILED test_rbac_white_cluster_endpoints.tavern.yaml::GET /cluster/{node_id}/configuration/{component}/{configuration}
FAILED test_rbac_white_cluster_endpoints.tavern.yaml::GET /cluster/{node_id}/info
FAILED test_rbac_white_cluster_endpoints.tavern.yaml::GET /cluster/{node_id}/logs
FAILED test_rbac_white_cluster_endpoints.tavern.yaml::GET /cluster/{node_id}/logs/summary
FAILED test_rbac_white_cluster_endpoints.tavern.yaml::GET /cluster/{node_id}/daemons/stats
FAILED test_rbac_white_cluster_endpoints.tavern.yaml::GET /cluster/{node_id}/stats
FAILED test_rbac_white_cluster_endpoints.tavern.yaml::GET /cluster/{node_id}/status
FAILED test_rbac_white_cluster_endpoints.tavern.yaml::GET /cluster/api/config
FAILED test_rbac_white_cluster_endpoints.tavern.yaml::PUT /cluster/restart
FAILED test_rbac_white_cluster_endpoints.tavern.yaml::PUT /cluster/{node_id}/configuration
============================================================================ 19 failed, 25 passed, 22 warnings in 719.38s (0:11:59) =============================================================================
```
### COMPLETE LOG

[complete log.txt](https://github.com/wazuh/wazuh/files/13229274/complete.log.txt)

**Comment by mhamra** [comment]
### UPDATE
All AITs passed OK in this [check](https://github.com/wazuh/wazuh/pull/19712/checks?check_run_id=18269348125).

File: framework/wazuh/core/wazuh_socket.py [hunk 1]
```diff
@@ -129,68 +128,49 @@ async def connect(self, path_to_socket: str):
...
-        return data
 
-    async def receive(self, header_size: int = None) -> bytes:
+        try:
+            self.writer.write(pack(header_format, len(msg_bytes)) + msg_bytes)
+            await self.writer.drain()
+        except (ConnectionResetError, OSError) as exc:
+            raise WazuhException(1014, "Socket connection was closed") from exc
+
+    async def receive(self, header_format:str="<I", header_size:int=4) -> bytes:
```
Comment (GGP1) [inline framework/wazuh/core/wazuh_socket.py:166]:
```suggestion
    async def receive(self, header_format: str ="<I", header_size: int =4) -> bytes:
```

File: framework/wazuh/stats.py [hunk 2]
```diff
@@ -219,7 +223,8 @@ def get_daemons_stats(daemons_list: list = None) -> AffectedItemsWazuhResult:
 
     for daemon in daemons_list or daemon_socket_mapping.keys():
         try:
-            result.affected_items.append(get_daemons_stats_socket(daemon_socket_mapping[daemon]))
+            result.affected_items.append(
+                await get_daemons_stats_socket(daemon_socket_mapping[daemon]))
```
Comment (GGP1) [inline framework/wazuh/stats.py:227]:
```suggestion
            res = await get_daemons_stats_socket(daemon_socket_mapping[daemon])
            result.affected_items.append(res)
```

File: framework/wazuh/core/wazuh_socket.py [hunk 3]
```diff
@@ -233,9 +212,9 @@ async def send(self, msg: str, header_format: str = None) -> bytes:
         bytes
             Bytes sent.
         """
-        return await WazuhAsyncSocket.send(self, dumps(msg).encode(), header_format)
+        return await super().send(msg_bytes=dumps(msg_bytes).encode(), header_format=header_format)
 
-    async def receive(self, header_size: int = None) -> dict:
+    async def receive_json(self, header_format="<I", header_size=4) -> dict:
```
Comment (GGP1) [inline framework/wazuh/core/wazuh_socket.py:217]:
```suggestion
    async def receive_json(self, header_format: str ="<I", header_size: int =4) -> dict:
```

**Review by GGP1 — CHANGES_REQUESTED** [review]

**Review by GGP1 — DISMISSED** [review]
LGTM

**Review by Selutario — CHANGES_REQUESTED** [review]
Looks good! I also checked whether the cluster `sendsync` protocol keeps working as expected, and it does, although `await` should be removed here:
https://github.com/wazuh/wazuh/blob/ad3fb60669a65beb1e3b96229960b38c1271532a/framework/wazuh/core/wazuh_socket.py#L270

Only some minor changes were requested additionally. It will be necessary to rebase with the master branch.

File: framework/wazuh/core/wazuh_socket.py [hunk 4]
```diff
@@ -129,68 +128,49 @@ async def connect(self, path_to_socket: str):
...
-        return data
 
-    async def receive(self, header_size: int = None) -> bytes:
+        try:
+            self.writer.write(pack(header_format, len(msg_bytes)) + msg_bytes)
+            await self.writer.drain()
+        except (ConnectionResetError, OSError) as exc:
+            raise WazuhException(1014, "Socket connection was closed") from exc
+
+    async def receive(self, header_format: str ="<I", header_size: int =4) -> bytes:
```
[the remark this answers is not in the document]
Comment (mhamra) [reply framework/wazuh/core/wazuh_socket.py]: Change committed:

https://github.com/wazuh/wazuh/blob/67860bcbcbc24331f817037c9f4886e335cde97a/framework/wazuh/core/wazuh_socket.py#L167

File: framework/wazuh/core/wazuh_socket.py [hunk 5]
```diff
@@ -205,21 +185,20 @@ async def receive(self, header_size: int = None) -> bytes:
...
-    """Handler class to connect and operate asynchronously with a socket using messages in JSON format."""
+    """Handler class to connect and operate asynchronously with a socket using 
+    messages in JSON format."""
 
-    def __init__(self):
-        WazuhAsyncSocket.__init__(self)
-
-    async def send(self, msg: str, header_format: str = None) -> bytes:
-        """Converts the message from JSON format to bytes and send it to the socket. Returns that message.
+    async def send(self, msg_bytes: str, header_format: str ="<I") -> bytes:
```
[the remark this answers is not in the document]
Comment (mhamra) [reply framework/wazuh/core/wazuh_socket.py]: Changes committed:

https://github.com/wazuh/wazuh/blob/67860bcbcbc24331f817037c9f4886e335cde97a/framework/wazuh/core/wazuh_socket.py#L200

File: framework/wazuh/core/wazuh_socket.py [hunk 6]
```diff
@@ -205,21 +185,20 @@ async def receive(self, header_size: int = None) -> bytes:
...
+        """Converts the message from JSON format to bytes and send it to the socket. 
+        Returns that message.
```
[the remark this answers is not in the document]
Comment (mhamra) [reply framework/wazuh/core/wazuh_socket.py]: Change committed:

https://github.com/wazuh/wazuh/blob/67860bcbcbc24331f817037c9f4886e335cde97a/framework/wazuh/core/wazuh_socket.py#L201

**Comment by mhamra** [comment]
### UPDATE
- The branch was rebased to master.
- All the requested changes are committed to the branch.

**Review by Selutario — CHANGES_REQUESTED** [review]
I'd like this to be released in [4.9.0](https://github.com/wazuh/wazuh/issues/19394#event-a40ed22e-1071-58fd-8add-c0072f67e319) so we can see the behavior in prod before refactoring all the framework. Since master is currently 5.0.0, could you rebase to 4.9.0 branch instead? Everything else looks good!

**Review by Selutario — APPROVED** [review]
I finally decided that it's okay to merge this in 5.0. The failing check is unrelated to these changes, and should be fixed once the changed applied here reach the master branch:
- https://github.com/wazuh/wazuh/issues/21822

**Pull request merged**